src/txn/txn_recover.c

```diff
@@ -841,11 +841,19 @@ __wt_txn_recover(WT_SESSION_IMPL *session, const char *cfg[])
         WT_ERR(ret);
     }
 
-    /* Check whether the history store exists. */
-    WT_ERR(__hs_exists(session, metac, cfg, &hs_exists));
-
     /* Scan the metadata to find the live files and their IDs. */
     WT_ERR(__recovery_file_scan(&r));
+
+    /*
```

Comment:
This is the fix for the issue that I've been chasing.

The `conn->base_write_gen` is updated as we scan the metadata file in `__recovery_file_scan`. So the value will jump within this function call.

Checking with `__hs_exists` BEFORE that happens will open the history store btree and calculate the `btree->base_write_gen` with the wrong `conn->base_write_gen`. If we leave it until after, `conn->base_write_gen` gets populated correctly and we don't have this issue.

It explains a lot of side effects because history store pages that get written during the first time that the history store dhandle is open will have low write gens (below the `conn->base_write_gen`). If we re-open the history store tree, those pages will appear to have been written in a previous run.